### PR TITLE
fix(consensus): retry transient L2 block visibility misses

### PR DESCRIPTION
**Pull request by niran**
## Context

While investigating duplicate blobs on Base Sepolia, we found base-consensus repeatedly logging `Pipeline reset error while preparing payload attributes` and `BlockNotFound` for an L2 parent block that the local execution client had just imported. The source of L2 RPC data was the colocated execution RPC, and the execution logs showed the block becoming canonical shortly after the failed read.

The race appears to be in the local execution RPC visibility path: a block can be added to the canonical chain before `eth_getBlockByNumber` can read it through the RPC cache or committed database path. During that brief window, the RPC returns `null` even though the block is not durably missing.

OP Node handles the analogous parent system-config miss as temporary. `PreparePayloadAttributes` calls `SystemConfigByL2Hash`, wraps errors as temporary, and lets the sequencer retry instead of treating the miss as a reset.

## Change

This PR adds a small retry budget around `AlloyL2ChainProvider::block_by_number` when the RPC returns `None` for a block. The retry is scoped to nil block responses only:

- cache hits are unchanged
- successful reads are cached as before
- transport errors still return immediately as transport errors
- durable misses still return `BlockNotFound` after the retry budget is exhausted

Each retry remains counted as an L2 chain request because it is a real RPC round-trip. A dedicated `l2_block_visibility_retries` counter tracks the visibility-race retry path separately so operators can observe how often this happens without inferring it from request volume.

This keeps the existing reset semantics for blocks that remain missing, while absorbing the local execution RPC visibility race before it is surfaced as a reset-class error.

## Tests

- `cargo test -p base-consensus-providers`
- `cargo clippy -p base-consensus-providers --all-targets -- -D warnings`

type=routine
risk=low
impact=sev5

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: crates/consensus/providers/src/l2_chain_provider.rs [hunk 1]
```diff
@@ -183,6 +187,62 @@ impl AlloyL2ChainProvider {
...
+        inner.get_block_by_number(number.into()).full().await
+    })
+    .map_err(|e| {
+        Metrics::l2_chain_errors(L2_BLOCK_REF_BY_NUMBER_METHOD).increment(1);
+        AlloyL2ChainProviderError::Transport(e)
+    })
+    .map(|block| {
+        block.map(|block| block.into_consensus().map_transactions(|t| t.inner.inner.into_inner()))
+    })
+}
```
[the remark this answers is not in the document]
Comment (refcell) [reply crates/consensus/providers/src/l2_chain_provider.rs]: This is a good callout. I'm not sure if we should record these retries as separate requests in the metric. I guess since we're actually retrying the request over the wire, I'd lean towards recording them individually.

[the remark this answers is not in the document]
Comment (niran) [reply crates/consensus/providers/src/l2_chain_provider.rs]: [24cbae1](https://github.com/base/base/pull/3009/commits/24cbae19822f3c47b64679fa07370ad9c11c3ca6)

**Comment by 0x00101010** [comment]
Nice fix. Two suggestions:

**1. Inline the loop into `block_by_number`.** The four-parameter generic seam (`F, Fut, Sleep, SleepFut`) and `Option<Duration>` indirection exist mainly to make the helper testable in isolation. Inlining drops the abstraction, removes the bare free functions, and reads top-to-bottom:

```rust
async fn block_by_number(&mut self, number: u64) -> Result<BaseBlock, Self::Error> {
    if let Some(block) = self.block_by_number_cache.get(&number) {
        return Ok(block.clone());
    }

    for attempt in 1..=L2_BLOCK_VISIBILITY_RETRY_ATTEMPTS {
        Metrics::l2_chain_requests(L2_BLOCK_REF_BY_NUMBER_METHOD).increment(1);

        let block = base_metrics::time!(
            Metrics::request_duration(L2_BLOCK_REF_BY_NUMBER_METHOD),
            { self.inner.get_block_by_number(number.into()).full().await }
        )
        .map_err(|e| {
            Metrics::l2_chain_errors(L2_BLOCK_REF_BY_NUMBER_METHOD).increment(1);
            AlloyL2ChainProviderError::Transport(e)
        })?;

        if let Some(block) = block {
            let block = block.into_consensus().map_transactions(|t| t.inner.inner.into_inner());
            self.block_by_number_cache.put(number, block.clone());
            return Ok(block);
        }

        if attempt < L2_BLOCK_VISIBILITY_RETRY_ATTEMPTS {
            Metrics::l2_block_visibility_retries().increment(1);
            tracing::debug!(block_number = number, attempt, "L2 block not visible yet; retrying");
            tokio::time::sleep(L2_BLOCK_VISIBILITY_RETRY_DELAY).await;
        }
    }

    Err(AlloyL2ChainProviderError::BlockNotFound(number))
}
```

**2. Test coverage.** The current unit tests exercise the helper but not the real `AlloyL2ChainProvider::block_by_number` wiring (`RootProvider` clone, JSON-RPC `null` decoding, transport-error short-circuit). With the loop inlined, please add one `httpmock`-based test through the real provider with `#[tokio::test(start_paused = true)]` covering: null→null→block succeeds; all-null returns `BlockNotFound`; transport error returns immediately without retry.

**Review by 0x00101010 — APPROVED** [review]
